onnx_tf/handlers/backend/pad.py

```diff
@@ -17,40 +16,63 @@ def _common(cls, node, **kwargs):
     num_dim = len(tensor_dict[node.inputs[0]].get_shape())
     mode = node.attrs.pop("mode", "constant")
 
+    def check_positive(pads):
+      p = tf.greater_equal(pads, tf.zeros((1), dtype=pads.dtype))
+      r = tf.reduce_all(p)
+      return r
+
+    def process_neg_pads(x, paddings):
+      # Process negative paddings differently since tf.pad
+      # doesn't support negative paddings
+      # The ONNX logic is similar to tf.slice. So we just
+      # need to compute the begins and sizes for slice op
+
+      i_shape = tf.shape(x, out_type=paddings.dtype)
+      i_rank = tf.cast(tf.rank(x), paddings.dtype)
+      begins = tf.negative(tf.gather(paddings, tf.range(i_rank)))
+      ends = i_shape + tf.gather(paddings, tf.range(i_rank, i_rank*2))
+      sizes = ends - begins
+      result=tf.slice(x, begins, sizes)
+      return [result]
+
+    def process_pos_pads(x, paddings):
+
+      def _symmetric_pad(i, x):
+        paddings_i = tf.map_fn(lambda e: tf.where(i < e, 1, 0), paddings)
+        paddings_i = tf.reshape(paddings_i, [num_dim, 2])
+        x = tf.pad(x, paddings_i, 'SYMMETRIC')
+        return i + 1, x
+
+      # tf requires int32 paddings
+      paddings = tf.cast(tf.transpose(tf.reshape(paddings, [2, num_dim])),
+                         dtype=tf.int32)
+
+      if mode.lower() == "edge":
+        paddings = tf.reshape(paddings, [-1])
+        max_i = tf.reduce_max(paddings)
+        _, x = tf.while_loop(
```

Comment:
Hi Chin, I either forgot or never understood what this part is doing, could you give me some hints about what's going on with the while loop? Some comments may help with future reviews, too.

Reply:
The while loop code is not changed in this PR, just added with indentation. It does look a bit strange. I will need to investigate what it is doing and whether it works as expected.

Reply:
Okay, now I understand the while loop is to replace the python implementation of edge mode, np.pad(x, pads, mode="edge"). Basically it goes through the tf.pad 'SYMMETRIC' mode to pad one value at a time for both sides and all dimensions. I added some comments and hope it helps.